FastSimulation/TrackerSimHitProducer/plugins/TrackerSimHitProducer.cc

```diff
@@ -0,0 +1,246 @@
+    		  if(hitPair.second){
+    		  	distAndHits.insert(distAndHits.end(), hitPair);
+    		  }
+    	    }
+    	}
+    }
+
+    // Fill simHitContainer
+    for(std::map<double, PSimHit*>::const_iterator it = distAndHits.begin(); it != distAndHits.end(); it++){
+    	simHitContainer_->push_back(*(it->second));
+    }
+    
+}
+
+// Also returns distance to simHit since hits have to be ordered (in time) afterwards. Necessary to do explicit copy of TrajectoryStateOnSurface particle (not call by reference)
+std::pair<double, PSimHit*> fastsim::TrackerSimHitProducer::createHitOnDetector(const TrajectoryStateOnSurface & particle, int pdgId, int simTrackId, const GeomDet & detector, GlobalPoint & refPos)
+{
+    //
+    // determine position and momentum of particle in the coordinate system of the detector
+    //
+    LocalPoint localPosition;
+    LocalVector localMomentum;
+    // if the particle is close enough, no further propagation is needed
+    if ( fabs( detector.toLocal(particle.globalPosition()).z()) < onSurfaceTolerance_) 
+    {
+	   localPosition = particle.localPosition();
+	   localMomentum = particle.localMomentum();
+    }
+    // else, propagate 
+    else 
+    {
+    	// find crossing of particle with 
+    	HelixArbitraryPlaneCrossing crossing(particle.globalPosition().basicVector(),
+    					      particle.globalMomentum().basicVector(),
+    					      particle.transverseCurvature(),
+    					      anyDirection);
+    	std::pair<bool,double> path = crossing.pathLength(detector.surface());
+    	// case propagation succeeds
+    	if (path.first) 	
+    	{
+    	    localPosition = detector.toLocal( GlobalPoint( crossing.position(path.second)));
+    	    localMomentum = detector.toLocal( GlobalVector( crossing.direction(path.second)));
+    	    localMomentum = localMomentum.unit() * particle.localMomentum().mag();
+    	}
+    	// case propagation fails
+    	else
+    	{
+    	    return std::pair<double, PSimHit*>(0, 0);
+    	}
+    }
+
+    // 
+    // find entry and exit point of particle in detector
+    //
+    const Plane& detectorPlane = detector.surface();
+    float halfThick = 0.5*detectorPlane.bounds().thickness();
+    float pZ = localMomentum.z();
+    LocalPoint entry = localPosition + (-halfThick/pZ) * localMomentum;
+    LocalPoint exit = localPosition + halfThick/pZ * localMomentum;
+    float tof = particle.globalPosition().mag() / 29.9792458 ; // in nanoseconds
+    
+    //
+    // make sure the simhit is physically on the module
+    //
+    double boundX = detectorPlane.bounds().width()/2.;
+    double boundY = detectorPlane.bounds().length()/2.;
+    // Special treatment for TID and TEC trapeziodal modules
+    unsigned subdet = DetId(detector.geographicalId()).subdetId(); 
+    if ( subdet == 4 || subdet == 6 ) 
+	boundX *=  1. - localPosition.y()/detectorPlane.position().perp();
```

Comment:
In the 1st file of this PR speed of light is added as a const, in this class this const is hard-coded. I would left one or another.